Framework/MDAlgorithms/src/IntegratePeaksMDHKL.cpp

```diff
@@ -0,0 +1,319 @@
+      make_unique<WorkspaceProperty<PeaksWorkspace>>("OutputWorkspace", "",
+                                                     Direction::Output),
+      "The output PeaksWorkspace will be a copy of the input PeaksWorkspace "
+      "with the peaks' integrated intensities.");
+}
+
+//----------------------------------------------------------------------------------------------
+/**
+ * Execute the algorithm.
+ */
+void IntegratePeaksMDHKL::exec() {
+  IMDWorkspace_sptr m_inputWS = getProperty("InputWorkspace");
+  Mantid::DataObjects::MDFramesToSpecialCoordinateSystem converter;
+  boost::optional<Mantid::Kernel::SpecialCoordinateSystem> coordinateSystem =
+      converter(m_inputWS.get());
+  if (*coordinateSystem != Mantid::Kernel::SpecialCoordinateSystem::HKL) {
+    std::stringstream errmsg;
+    errmsg << "Input MDWorkspace's coordinate system is not HKL.";
+    throw std::invalid_argument(errmsg.str());
+  }
+
+  /// Peak workspace to integrate
+  PeaksWorkspace_sptr inPeakWS = getProperty("PeaksWorkspace");
+  const double box = getProperty("DeltaHKL");
+  const int gridPts = getProperty("GridPoints");
+  const int neighborPts = getProperty("NeighborPoints");
+  /// Output peaks workspace, create if needed
+  PeaksWorkspace_sptr peakWS = getProperty("OutputWorkspace");
+  if (peakWS != inPeakWS)
+    peakWS = inPeakWS->clone();
+
+  MatrixWorkspace_sptr flux = getProperty("FluxWorkspace");
+  MatrixWorkspace_sptr sa = getProperty("SolidAngleWorkspace");
+
+  IMDEventWorkspace_sptr m_eventWS =
+      boost::dynamic_pointer_cast<IMDEventWorkspace>(m_inputWS);
+  IMDHistoWorkspace_sptr m_histoWS =
+      boost::dynamic_pointer_cast<IMDHistoWorkspace>(m_inputWS);
+  int npeaks = peakWS->getNumberPeaks();
+
+  auto prog = make_unique<Progress>(this, 0.3, 1.0, npeaks);
+  PARALLEL_FOR1(peakWS)
+  for (int i = 0; i < npeaks; i++) {
+    PARALLEL_START_INTERUPT_REGION
+
+    IPeak &p = peakWS->getPeak(i);
+    // round to integer
+    int h = static_cast<int>(std::round(p.getH()));
+    int k = static_cast<int>(std::round(p.getK()));
+    int l = static_cast<int>(std::round(p.getL()));
+    MDHistoWorkspace_sptr histoBox;
+    if (m_histoWS) {
+      histoBox = cropHisto(h, k, l, box, m_histoWS);
+    } else if (sa && flux) {
+      histoBox = normalize(h, k, l, box, gridPts, flux, sa, m_eventWS);
+    } else {
+      histoBox = binEvent(h, k, l, box, gridPts, m_eventWS);
+    }
+    double intensity = 0.0;
+    double errorSquared = 0.0;
+    integratePeak(neighborPts, histoBox, intensity, errorSquared);
+    p.setIntensity(intensity);
+    p.setSigmaIntensity(sqrt(errorSquared));
+    prog->report();
+    PARALLEL_END_INTERUPT_REGION
+  }
+  PARALLEL_CHECK_INTERUPT_REGION
+  // Save the output
+  setProperty("OutputWorkspace", peakWS);
+}
+
+MDHistoWorkspace_sptr IntegratePeaksMDHKL::normalize(
+    int h, int k, int l, double box, int gridPts, MatrixWorkspace_sptr flux,
+    MatrixWorkspace_sptr sa, IMDEventWorkspace_sptr ws) {
```

Comment:
Since the `IPropertyManager` takes `ws` by const reference, should we also pass it here by const reference?